File: backend/src/main/java/com/suplr/backend/controller/AdminController.java
package com.suplr.backend.controller;

import com.suplr.backend.dto.AdminDtos.*;
import com.suplr.backend.dto.AuthDtos.SupplierResponse;
import com.suplr.backend.entity.Supplier;
import com.suplr.backend.repository.SupplierRepository;
import com.suplr.backend.security.JwtService;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import com.suplr.backend.service.EmailService;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.List;

@Slf4j
@RestController
@RequestMapping("/admin")
@PreAuthorize("hasRole('ADMIN')")
@RequiredArgsConstructor
public class AdminController {

    private final SupplierRepository supplierRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final EmailService emailService;

    @PersistenceContext
    private EntityManager em;


    @PostMapping("/suppliers")
    @ResponseStatus(HttpStatus.CREATED)
    @Transactional
    public SupplierResponse createSupplier(@Valid @RequestBody CreateSupplierRequest req) {
        if (supplierRepository.existsByEmail(req.email())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email already registered");
        }
        String rawPassword = req.password();
        Supplier supplier = Supplier.builder()
                .name(req.name())
                .email(req.email())
                .passwordHash(passwordEncoder.encode(rawPassword))
                .plan("pro")
                .mustChangePassword(true)
                .build();
        supplier = supplierRepository.save(supplier);
        emailService.sendWelcomeEmail(req.name(), req.email(), rawPassword);
        return SupplierResponse.from(supplier);
    }


    @GetMapping("/suppliers")
    public List<SupplierWithStatsResponse> listSuppliers() {
        OffsetDateTime sevenDaysAgo = OffsetDateTime.now(ZoneOffset.UTC).minusDays(7);

        @SuppressWarnings("unchecked")
        List<Object[]> rows = em.createNativeQuery("""
                SELECT s.id,
                       COUNT(DISTINCT o.id)  AS order_count,
                       COUNT(DISTINCT i.id)  AS invoice_count,
                       COUNT(DISTINCT c.id)  AS client_count,
                       MAX(o.created_at)     AS last_order_at
                FROM suppliers s
                LEFT JOIN orders   o ON o.supplier_id = s.id
                LEFT JOIN invoices i ON i.supplier_id = s.id
                LEFT JOIN clients  c ON c.supplier_id = s.id
                GROUP BY s.id
                ORDER BY s.created_at DESC
                """).getResultList();

        return rows.stream().map(r -> {
            Integer sid = ((Number) r[0]).intValue();
            Supplier s = supplierRepository.findById(sid)
                    .orElseThrow();
            int orders = ((Number) r[1]).intValue();
            int invoices = ((Number) r[2]).intValue();
            int clients = ((Number) r[3]).intValue();
            boolean active = r[4] != null &&
                    toODT(r[4]).isAfter(sevenDaysAgo);
            return SupplierWithStatsResponse.from(s, orders, invoices, clients, active);
        }).toList();
    }


    @PatchMapping("/suppliers/{supplierId}/plan")
    @Transactional
    public SupplierResponse setPlan(
            @PathVariable Integer supplierId,
            @Valid @RequestBody PlanRequest req
    ) {
        if (!"pro".equals(req.plan())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Plan must be 'pro'");
        }
        Supplier supplier = getSupplier(supplierId);
        supplier.setPlan(req.plan());
        return SupplierResponse.from(supplierRepository.save(supplier));
    }


    @PatchMapping("/suppliers/{supplierId}/suspend")
    @Transactional
    public SupplierResponse toggleSuspend(@PathVariable Integer supplierId) {
        Supplier supplier = getSupplier(supplierId);
        if (supplier.isAdmin()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot suspend an admin account");
        }
        supplier.setSuspended(!supplier.isSuspended());
        return SupplierResponse.from(supplierRepository.save(supplier));
    }


    @DeleteMapping("/suppliers/{supplierId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @Transactional
    public void deleteSupplier(@PathVariable Integer supplierId) {
        Supplier supplier = getSupplier(supplierId);
        if (supplier.isAdmin()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot delete an admin account");
        }
        em.createNativeQuery("DELETE FROM invoices WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM order_items WHERE order_id IN (SELECT id FROM orders WHERE supplier_id = :sid)")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM orders WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM messages WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM pending_orders WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM clients WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM products WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM whatsapp_connections WHERE supplier_id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
        em.createNativeQuery("DELETE FROM suppliers WHERE id = :sid")
                .setParameter("sid", supplierId).executeUpdate();
    }

    @PostMapping("/suppliers/{supplierId}/impersonate")
    public ImpersonateResponse impersonate(@PathVariable Integer supplierId) {
        Supplier supplier = getSupplier(supplierId);
        if (supplier.isAdmin()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot impersonate admin");
        }
        return new ImpersonateResponse(jwtService.generateToken(supplierId));
    }

    @GetMapping("/orders")
    public List<AdminOrderRow> listAllOrders() {
        @SuppressWarnings("unchecked")
        List<Object[]> rows = em.createNativeQuery("""
                SELECT o.id, o.status, o.created_at,
                       s.name AS supplier_name,
                       c.name AS client_name
                FROM orders o
                JOIN suppliers s ON s.id = o.supplier_id
                JOIN clients   c ON c.id = o.client_id
                ORDER BY o.created_at DESC
                LIMIT 200
                """).getResultList();

        return rows.stream().map(r -> new AdminOrderRow(
                ((Number) r[0]).intValue(),
                (String) r[1],
                toODT(r[2]),
                (String) r[3],
                (String) r[4]
        )).toList();
    }

    @PostMapping("/broadcast")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @Transactional
    public void broadcast(@Valid @RequestBody BroadcastRequest req) {
        List<Supplier> suppliers = supplierRepository
                .findByIsAdminFalseAndSuspendedFalse();
        if (suppliers.isEmpty()) return;
        for (Supplier s : suppliers) {
            emailService.sendBroadcastEmail(s.getEmail(), req.subject(), req.message());
        }
    }


    private Supplier getSupplier(Integer id) {
        return supplierRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Supplier not found"));
    }

    private OffsetDateTime toODT(Object o) {
        if (o instanceof java.sql.Timestamp ts)
            return ts.toInstant().atOffset(java.time.ZoneOffset.UTC);
        if (o instanceof java.time.LocalDateTime ldt)
            return ldt.atOffset(java.time.ZoneOffset.UTC);
        return null;
    }
}

File: backend/src/admin/admin.controller.ts
import {
  Controller,
  Get,
  Post,
  Patch,
  Delete,
  Param,
  Body,
  UseGuards,
  ParseIntPipe,
  HttpCode,
  HttpStatus,
  NotFoundException,
  BadRequestException,
} from '@nestjs/common';
import { InjectRepository, InjectDataSource } from '@nestjs/typeorm';
import { Repository, DataSource } from 'typeorm';
import { IsString, IsEmail, MinLength } from 'class-validator';
import { JwtAuthGuard } from '../common/guards/jwt-auth.guard.js';
import { AdminGuard } from '../common/guards/admin.guard.js';
import { Supplier } from '../entities/supplier.entity.js';
import { EmailService } from '../email/email.service.js';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';

class CreateSupplierRequest {
  @IsString() name!: string;
  @IsEmail() email!: string;
  @MinLength(8) password!: string;
}
class PlanRequest {
  @IsString() plan!: string;
}
class AdminBroadcastRequest {
  @IsString() subject!: string;
  @IsString() message!: string;
}

@UseGuards(JwtAuthGuard, AdminGuard)
@Controller('admin')
export class AdminController {
  constructor(
    @InjectRepository(Supplier)
    private readonly supplierRepo: Repository<Supplier>,
    @InjectDataSource() private readonly ds: DataSource,
    private readonly emailService: EmailService,
    private readonly jwtService: JwtService,
  ) {}

  @Post('suppliers')
  @HttpCode(HttpStatus.CREATED)
  async createSupplier(@Body() req: CreateSupplierRequest) {
    const exists = await this.supplierRepo.findOne({
      where: { email: req.email },
    });
    if (exists) throw new BadRequestException('Email already registered');
    const hash = await bcrypt.hash(req.password, 12);
    const supplier = await this.supplierRepo.save(
      this.supplierRepo.create({
        name: req.name,
        email: req.email,
        passwordHash: hash,
        plan: 'pro',
        mustChangePassword: true,
      }),
    );
    await this.emailService.sendWelcomeEmail(req.name, req.email, req.password);
    return this.toResponse(supplier);
  }

  @Get('suppliers')
  async listSuppliers() {
    const rows = (await this.ds.query(`
      SELECT s.id, COUNT(DISTINCT o.id)::int AS order_count,
             COUNT(DISTINCT i.id)::int AS invoice_count,
             COUNT(DISTINCT c.id)::int AS client_count,
             MAX(o.created_at) AS last_order_at
      FROM suppliers s
      LEFT JOIN orders o ON o.supplier_id = s.id
      LEFT JOIN invoices i ON i.supplier_id = s.id
      LEFT JOIN clients c ON c.supplier_id = s.id
      GROUP BY s.id ORDER BY s.created_at DESC
    `)) as {
      id: number;
      order_count: number;
      invoice_count: number;
      client_count: number;
      last_order_at: Date | null;
    }[];
    const sevenDaysAgo = new Date(Date.now() - 7 * 86400000);
    return Promise.all(
      rows.map(async (r) => {
        const s = await this.supplierRepo.findOne({ where: { id: r.id } });
        if (!s) return null;
        return {
          ...this.toResponse(s),
          orderCount: r.order_count,
          invoiceCount: r.invoice_count,
          clientCount: r.client_count,
          active:
            r.last_order_at != null && new Date(r.last_order_at) > sevenDaysAgo,
        };
      }),
    );
  }

  @Patch('suppliers/:supplierId/plan')
  async setPlan(
    @Param('supplierId', ParseIntPipe) supplierId: number,
    @Body() req: PlanRequest,
  ) {
    if (req.plan !== 'pro') throw new BadRequestException('Plan must be pro');
    const s = await this.getSupplier(supplierId);
    s.plan = req.plan;
    return this.toResponse(await this.supplierRepo.save(s));
  }

  @Patch('suppliers/:supplierId/suspend')
  async toggleSuspend(@Param('supplierId', ParseIntPipe) supplierId: number) {
    const s = await this.getSupplier(supplierId);
    if (s.isAdmin)
      throw new BadRequestException('Cannot suspend an admin account');
    s.suspended = !s.suspended;
    return this.toResponse(await this.supplierRepo.save(s));
  }

  @Delete('suppliers/:supplierId')
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteSupplier(@Param('supplierId', ParseIntPipe) supplierId: number) {
    const s = await this.getSupplier(supplierId);
    if (s.isAdmin)
      throw new BadRequestException('Cannot delete an admin account');
    await this.ds.transaction(async (em) => {
      const q = (sql: string) => em.query(sql, [supplierId]);
      // invoices and order_items must precede orders (FK)
      await Promise.all([
        q('DELETE FROM invoices WHERE supplier_id = $1'),
        q(
          'DELETE FROM order_items WHERE order_id IN (SELECT id FROM orders WHERE supplier_id = $1)',
        ),
      ]);
      await q('DELETE FROM orders WHERE supplier_id = $1');
      await Promise.all([
        q('DELETE FROM messages WHERE supplier_id = $1'),
        q('DELETE FROM pending_orders WHERE supplier_id = $1'),
        q('DELETE FROM clients WHERE supplier_id = $1'),
        q('DELETE FROM products WHERE supplier_id = $1'),
        q('DELETE FROM whatsapp_connections WHERE supplier_id = $1'),
      ]);
      await em.query('DELETE FROM suppliers WHERE id = $1', [supplierId]);
    });
  }

  @Post('suppliers/:supplierId/impersonate')
  async impersonate(@Param('supplierId', ParseIntPipe) supplierId: number) {
    const s = await this.getSupplier(supplierId);
    if (s.isAdmin) throw new BadRequestException('Cannot impersonate admin');
    const token = await this.jwtService.signAsync({ sub: supplierId });
    return { token };
  }

  @Get('orders')
  async listAllOrders() {
    const rows = (await this.ds.query(`
      SELECT o.id, o.status, o.created_at, s.name AS supplier_name, c.name AS client_name
      FROM orders o
      JOIN suppliers s ON s.id = o.supplier_id
      JOIN clients c ON c.id = o.client_id
      ORDER BY o.created_at DESC LIMIT 200
    `)) as {
      id: number;
      status: string;
      created_at: Date;
      supplier_name: string;
      client_name: string;
    }[];
    return rows.map((r) => ({
      id: r.id,
      status: r.status,
      createdAt: r.created_at,
      supplierName: r.supplier_name,
      clientName: r.client_name,
    }));
  }

  @Post('broadcast')
  @HttpCode(HttpStatus.NO_CONTENT)
  async broadcast(@Body() req: AdminBroadcastRequest) {
    const suppliers = await this.supplierRepo.find({
      where: { isAdmin: false, suspended: false },
    });
    await Promise.allSettled(
      suppliers.map((s) =>
        this.emailService.sendBroadcastEmail(s.email, req.subject, req.message),
      ),
    );
  }

  private async getSupplier(id: number): Promise<Supplier> {
    const s = await this.supplierRepo.findOne({ where: { id } });
    if (!s) throw new NotFoundException('Supplier not found');
    return s;
  }

  private toResponse(s: Supplier) {
    return {
      id: s.id,
      name: s.name,
      email: s.email,
      plan: s.plan,
      isAdmin: s.isAdmin,
      suspended: s.suspended,
      mustChangePassword: s.mustChangePassword,
      logo: s.logo,
      createdAt: s.createdAt,
    };
  }
}
